### ISSUE TITLE
jax.nn.softmax is inconsistent under jax.jit

### ISSUE THREAD

**Issue by francois-rozet** [jax-ml/jax#20856]
### Description

In the following code, `jax.nn.softmax` returns different results under `jax.jit`.

```python
import jax
import jax.numpy as jnp

x = jax.random.normal(jax.random.key(0), (3, 1, 1))

def f(x):
    return jax.nn.softmax(x, axis=0)

print(f(x))
print(jax.jit(f)(x))
```

```
[[[0.75250584]]

 [[0.0755428 ]]

 [[0.17195134]]]
[[[1.]]

 [[1.]]

 [[1.]]]
```

Using a "numerically safe" version of `softmax` based on `log_softmax` solves the issue.

```python
def f(x):
    return jnp.exp(jax.nn.log_softmax(x, axis=0))

print(f(x))
print(jax.jit(f)(x))
```

```
[[[0.75250584]]

 [[0.07554279]]

 [[0.17195135]]]
[[[0.75250584]]

 [[0.07554279]]

 [[0.17195135]]]
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.26.2
python: 3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:49:32)  [GCC 12.3.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='thinkpad', release='6.5.0-0.deb12.4-amd64', version='#1 SMP PREEMPT_DYNAMIC Debian 6.5.10-1~bpo12+1 (2023-11-23)', machine='x86_64')
```

**Comment by francois-rozet** [comment]
Note also that printing within `f` weirdly fixes the issue.

```python
import jax
import jax.numpy as jnp

x = jax.random.normal(jax.random.key(0), (3, 1, 1))

def f(x):
    jax.debug.print('{}', x[0, 0])
    return jax.nn.softmax(x, axis=0)

print(f(x))
print(jax.jit(f)(x))
```

```
[1.8160863]
[[[0.75250584]]

 [[0.0755428 ]]

 [[0.17195134]]]
[1.8160863]
[[[0.75250584]]

 [[0.0755428 ]]

 [[0.17195134]]]
```

**Comment by jakevdp** [comment]
Thanks for the report – it looks like this is a CPU-only issue; I get the expected result when running on GPU.

**Comment by jakevdp** [comment]
Also seems like it was introduced in JAX v0.4.22; JAX v0.4.21 and earlier returns the expected results for the handful of versions I've tried.

**Comment by francois-rozet** [comment]
> it looks like this is a CPU-only issue

Oh I should have checked. Anyway, it seems related to this comment:

https://github.com/google/jax/blob/e498bca2233fa427153f921f66aebd4df488aa9a/jax/_src/nn/functions.py#L536-L537

**Comment by jakevdp** [comment]
It looks like this bug appears only when the XLA softmax rewriter is enabled, so it's likely related to this XLA change: https://github.com/openxla/xla/pull/7540

**Comment by NeilGirdhar** [comment]
Whoa that is a mess.  Just looking at the onednn-softmax, it only supports `axis=-1` (it's hard-coded).

**Comment by vfdev-5** [comment]
@jakevdp I think this issue can be closed as solved as 0.4.30 gives same results from the repro code on CPU:
```
0.4.30
[[[0.75250584]]
 [[0.0755428 ]]
 [[0.17195134]]] {CpuDevice(id=0)}

[[[0.75250584]]
 [[0.07554279]]
 [[0.17195134]]] {CpuDevice(id=0)}
```
